### PR TITLE
EXT_lights_ies

### PR DESCRIPTION
**Pull request by tblutNV**
This extension adds support for referencing and storing IES light profiles and using them as point lights. IES light profile files are used as-is, similar to how PNG and JPEG images are used.

There are still some open questions regarding a valid MIME type for IES light profiles and if any additional properties are required to cover the needs of most users.

We don't want to force applications to fully support `KHR_lights_punctual` in order to implement this extension. Therefore, we choose to not extend `KHR_lights_punctual`, but rather make this a standalone extension. However, we are open for discussing the possibility to make extending punctual lights optional. The main benefit would be that spot lights could be enhanced with light profiles as well.
```javascript
{
    "nodes": [
        {
            "extensions": {
                "KHR_lights_punctual": {
                    "type": "point",
                    "color": [ 1.0, 0.8, 0.6 ],
                    "extensions": {
                        "KHR_lights_ies": {
                            "profile": 0,
                            "useProfileIntensity": true, // if "true" then ignore the light's intensity
                            "multiplier": 1.2, // non-negative intesity scale factor
                            "orientation": [0.707, 0, 0, 0.707] // rotation of the light profile
                        }
                    }
                }
            }
        }
    ]
}
```
An addition here could be the `useProfileIntensity` property to allow to either use the light profile's intensity or the punctual light's intensity.
In the case of using the punctual light's intensity the light profile's candela values must be scaled to the [0, 1] range: `candelaValues /= max(candelaValues)`. These values can then be multiplied by the light's `intensity` and `color`.

## Alternative Proposal for Storing IES Data

Based on the discussion and feedback received in the PBR TSG meeting on 8/29/2022, an alternative proposal is to store the relevant parts of the IES files directly in JSON. Metadata can be stored in the `extras` property with the keys being the field names as defined in the IES format specification in glTF-like camel case.
```javascript
{
    "extensions": {
        "KHR_lights_ies": {
            "lights": [
                {
                    "photometricType": 1, // 1: "Type C", 2: "Type B", 3: "Type A"
                    "candelaMultiplier": 1.0, // balast_factor * multiplier, must be applied to all candelaValues
                    "verticalAngles": [ 0, 22.5, 45, 67.5, 90 ], // in degrees
                    "horizontalAngles": [ 0, 45, 90 ], // in degrees, cc or ccw depending on photometricType
                    "candelaValues": [ // in candela (lm/sr), column-major (vertical values first)
                        100000, 50000, 25000, 10000, 5000,
                        100000, 35000, 16000, 8000, 3000,
                        100000, 20000, 10000, 5000, 1000
                    ],
                    "extras": {
                        "test": "Test ABC",
                        "testlab": "Lab ABC",
                        "manufac": "Manufacturer ABC"
                    }
                }
            ]
        }
    }
}
```
We omitted the TILT data as we don't believe that it is relevant for most rendering applications.

We're looking forward to further discussions!

### CODE REVIEW

**Comment by lexaknyazev** [comment]
With regards to the topic of storage, i.e., external/embedded ANSI/IES LM-63 data vs JSON-inlined parameters, I think we need to consider the following two issues.

#### ANSI/IES LM-63 Lifecycle
The extension spec currently does not require any specific revision of ANSI/IES LM-63. This is quite an important detail since ANSI/IES LM-63 files contain the version string inside. Supporting all 5 versions is probably not feasible while supporting only the latest would make lots of readily available ANSI/IES LM-63 files incompatible with this extension. The reasonable middle-ground may be supporting both 2019 and 2002 version strings while explicitly excluding external tilt files from the latter.

It's worth noting though that the latest (2019) revision says the following:
>This Standard will eventually be replaced by a new XML-based file format, ANSI/IES TM-33-18, Standard Format for the Electronic Transfer of Luminaire Optical Data (see Informative References). However, both Standards are valid until this transition is complete and this Standard has been deprecated.

We don't have a precedent of making and ratifying a new glTF extension that exclusively relies on a soon-to-be-deprecated external specification.

#### ANSI/IES LM-63 Scope
With external tilt files out of the way, there are still lots of ANSI/IES LM-63 fields that may or may not be relevant for glTF use. Normatively referencing an external spec generally means that implementations have to follow it even if some of its parts are not needed for glTF applications in practice. For example, ANSI/IES LM-63 data must always contain four required keywords (`TEST`, `TESTLAB`, `MANUFAC`, and `ISSUEDATE`), support both feet and meters, enforce max line length and CRLF line breaks, etc. Proper implementations would have to reject data not conforming to those rules even when deviations have no practical effect on glTF.

We should identify all glTF-relevant ANSI/IES LM-63 fields and keywords and evaluate whether providing them directly in JSON makes the new functionality more accessible for implementors. Alternatively, we should evaluate availability of open-source ANSI/IES LM-63 parser implementations, including test coverage for malformed inputs.

@tblutNV please comment on the applicability of the fields marked with `?` below.

| Field | ANSI/IES LM-63 | glTF JSON |
|---|:---:|:---:|
| Version | 2002 or 2019 | not needed |
| TEST | required | not needed |
| TESTLAB | required | not needed |
| TESTDATE | optional | not needed |
| NEARFIELD | optional | ? |
| MANUFAC | required | not needed |
| LUMCAT | optional | not needed |
| LUMINAIRE | optional | not needed |
| LAMPCAT | optional | not needed |
| LAMP | optional | not needed |
| BALLAST | optional | not needed |
| BALLASTCAT | optional | not needed |
| MAINTCAT | optional | not needed |
| DISTRIBUTION | optional | not needed |
| FLASHAREA | optional | ? |
| COLORCONSTANT | optional | ? |
| LAMPPOSITION | optional | ? |
| ISSUEDATE | required | not needed |
| FILEGENINFO | optional | not needed |
| OTHER | optional | not needed |
| SEARCH | optional | not needed |
| MORE | meta | not needed |
| TILT | NONE or INCLUDE | not needed |
| TILT: lamp to luminaire geometry | 1 or 2 or 3 | enum / not needed |
| TILT: number of tilt angles | integer | not needed |
| TILT: angles | list of numbers | array of numbers / not needed |
| TILT: multiplying factors | list of numbers | array of numbers / not needed |
| number of lamps | integer | ? |
| lumens per lamp | double | ? |
| candela multiplier | double | number |
| number of vertical angles | integer | not needed |
| number of horizontal angles | integer | not needed |
| photometric type | 1 or 2 or 3 | enum |
| units type | 1 or 2 | not needed |
| width | double | ? |
| length | double | ? |
| height | double | ? |
| ballast factor | double | ? |
| file generation type | integer | not needed |
| input watts | double | ? |
| vertical angles | list of numbers | array of numbers |
| horizontal angles | list of numbers | array of numbers |
| candela values | list of lists of numbers | flat array of numbers |

**Comment by lkettnerNV** [comment]
[lexaknyazev](https://github.com/lexaknyazev) W.r.t. to the IES fields, we ignore in fact all fields in the meta data, the TILT section, number of lamps, lumen per lamp, units type, width, length, height, and input watts. We respect the ballast factor (it is multiplied on top of the candela multiplier).  This should cover the questionmarks in the table.

A good observation is indeed that the IES standard does not include previous versions, most notably seen with the magic header. We would actually want to support all versions: IESNA91, IESNA:LM-63-1995, IESNA:LM-63-2002, IES:LM-63-2019.

**Comment by lkettnerNV** [comment]
More comments regarding a JSON encoding instead of referencing the IES file.

One could think of removing the photometric type altogether. They are basically just different orientations. In fact, we don't support type A, which is very rare (and can be converted to Type C in a suitable tool).

One could also think of reducing or eliminating the symmetry groups, basically unrolling the data to the full sphere (simplifying the use of the data at the cost of more storage).  That can work from out perspective.

However, resampling to a different, e.g., equidistant set of angles would not be a good idea. A renderer will do that typically too, so we get two re-samplings that do not know of each other, and it can harm the original data in ways that would make it not possible for us anymore to rely on the new encoding alone.

**Comment by lexaknyazev** [comment]
> We would actually want to support all versions: IESNA91, IESNA:LM-63-1995, IESNA:LM-63-2002, IES:LM-63-2019.

IIRC, there was also a 1986 version that did not have a version identifier inside at all. Supporting all of them would complicate implementations as there were a few breaking changes over time.

How would the extension normatively reference older revisions? Looks like they are not offered in the IES store.

**Comment by aidinabedi** [comment]
Hello! A quick curious question:

What's the reasoning behind introducing `orientation` in the schema?

At first glance it seems a bit redundant, when reading the following in the [overview](https://github.com/tblutNV/glTF/tree/KHR_lights_ies/extensions/2.0/Khronos/KHR_lights_ies#overview):
> The light profiles are referenced by nodes and inherit their transform. This also allows changing the light profile's orientation.

**Comment by lkettnerNV** [comment]
Indeed, we should adjust the wording in the description. We propose the additional orientation on the lightprofile since you want to orient it independently of the lumiere, there are for example ill oriented light profiles out there that you want to align.

**Comment by aidinabedi** [comment]
> We propose the additional orientation on the lightprofile since you want to orient it independently of the lumiere, there are for example ill oriented light profiles out there that you want to align.

If there are ill-oriented light profiles out there, shouldn't that be solved in the file itself? Seems strange that glTF should handle bad external files (regardless if it's a light profile, texture, or raw binary). [Unreal also supports IES](https://docs.unrealengine.com/5.0/en-US/using-ies-light-profiles-in-unreal-engine/) but doesn't provide additional orientation (our philosophy is "garbage in, garbage out").

**Comment by BruceCherniak** [comment]
I have begun a draft PR for the file-format independent json representation here https://github.com/KhronosGroup/glTF/pull/2221

**Comment by tblutNV** [comment]
I updated the extension name from KHR_lights_ies to EXT_lights_ies. The commits also include some minor additions, clarifications and implementation details based on the feedback we received.

File: extensions/2.0/Vendor/EXT_lights_ies/README.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+* [light.schema.json](schema/light.schema.json)
+
+## Known Implementations
+
+* [NVIDIA Omniverse](https://www.nvidia.com/en-us/omniverse/)
+
+## Resources
+
+* [IES LM-63-19 (IES Standard File Format for Electronic Transfer of Photometric Data and Related Information), IES Computer Committee](http://www.iesna.org)
+* [IES LM-75-01/R12 (Goniophotometer Types and Photometric Coordinates), IES Computer Commitee](http://www.iesna.org)
```
Comment (lexaknyazev) [inline extensions/2.0/Vendor/EXT_lights_ies/README.md:126]: The first link could directly point to the standard's page
- https://store.ies.org/product/lm-63-19-approved-method-ies-standard-file-format-for-the-electronic-transfer-of-photometric-data-and-related-information/.

The second reference should likely be (please confirm)
- https://store.ies.org/product/lm-75-19-approved-method-guide-to-goniometer-measurements-and-types-and-photometric-coordinate-systems/

File: extensions/2.0/Vendor/EXT_lights_ies/schema/lightProfile.schema.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{
+    "$schema": "http://json-schema.org/draft-04/schema",
+    "title": "light",
```
[the remark this answers is not in the document]
Comment (tblutNV) [reply extensions/2.0/Vendor/EXT_lights_ies/schema/lightProfile.schema.json]: Good catch. I missed that one.

File: extensions/2.0/Vendor/EXT_lights_ies/README.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+| Property | Description | Required |
+|:---------|:------------|:---------|
+| `uri` | The URI (or IRI) of the light profile. | No |
+| `bufferView` | The index of the bufferView that contains the IES light profile. | No |
+| `mimeType` | The light profile's media type. Must be `"application/x-ies-lm-63"`. | No |
+| `name` | glTF name of the light profile. | No, Default: `""` |
+
+## Adding Light Profile Instances to Nodes
+
+Light profiles can be attached to a node by defining the `extensions.EXT_lights_ies` property and, within that, the `light` property which defines an index into the `lights` array. Optionally, the `multiplier` and `color` properties can be defined to scale the light's intensity and tint the light, respectively. The RGB value of the `color` property is in linear space, clamped to the `[0, 1]` range, and acts as a per-wavelength multiplier. Note that values other than `1.0` for `multiplier` and `(1.0, 1.0, 1.0)` for `color` change the energy output of the light source. An energy preserving tint can be achieved by modifying `multiplier` to compensate for the energy loss: `multiplier *= 3 / dot(color, color)`.
```
[the remark this answers is not in the document]
Comment (tblutNV) [reply extensions/2.0/Vendor/EXT_lights_ies/README.md]: I agree, "per color channel" is more accurate here. Changed it.

**Review by lexaknyazev — APPROVED** [review]

**Review by emackey — APPROVED** [review]

**Comment by sebavan** [comment]
Sorry for necro posting, but I am wondering if there are any demo assets available ?

I am evaluating whether to implement the extension as I love the intent and it would be great to have samples I could rely on for the tests.

**Comment by deltakosh** [comment]
Hey team! I've just added support for IES profile in Babylon v8.0 and I would be more than happy to add support for EXT_light_ies.

Do we know a way to export that extension? (Blender does not seem to support it). Any sample file I can use to test?

**Comment by tblutNV** [comment]
Here is a sample file I made by hand for my own tests: [EXT_lights_ies_sample.zip](https://github.com/user-attachments/files/18015715/EXT_lights_ies_sample.zip)

This is how it should look like when rendered:
ext_lights_ies_example

Looking forward to seeing the extension implemented in Babylon!

**Comment by deltakosh** [comment]
@tblutNV do you have the same but with buffer and not external URL?
I've merged our support for IES so almost done to support the extension too

**Comment by tblutNV** [comment]
@deltakosh try these: [EXT_lights_ies_samples.zip](https://github.com/user-attachments/files/18065732/EXT_lights_ies_samples.zip)
One is with an embedded buffer and the other is with a data-URI.

**Comment by deltakosh** [comment]
Here we are: https://github.com/BabylonJS/Babylon.js/pull/15970

**Pull request merged**